### PR TITLE
Parameter to make line strength proportional to coupling strength in Plot

### PR DESCRIPTION
**Pull request by gjhuizing**
We added a parameter to make line strength proportional to coupling strength in Plot

### CODE REVIEW

File: src/ott/tools/plot.py [hunk 1]
```diff
@@ -120,13 +122,26 @@ def __call__(self, ot: Transport) -> List[plt.Artist]:
     cmap = plt.get_cmap(self._cmap)
     self._lines = []
     for start, end, strength in lines:
+
+      # Make the transparency of the lines depend on the strength of the coupling.
+      alpha = 0.7
```
Comment (michalk8) [inline src/ott/tools/plot.py:127]: I'd consider these changes:
- move the computation of normalization out of the loop (it's costly)
- allow passing `alpha: float = 0.7` in the `__init__` instead of it being hard-coded
- assign `ot.matrix` to a variable (it it being possibly re-computed)
- also to save some computation

Comment (gjhuizing) [reply src/ott/tools/plot.py]: Adressed these changes

- I put the max and min out of the loop, but I kept a quick normalization step in the loop (scalar multiplication only)
- I put alpha in the init
- I assigned ot.matrix to a variable

File: src/ott/tools/plot.py [hunk 2]
```diff
@@ -119,14 +123,34 @@ def __call__(self, ot: Transport) -> List[plt.Artist]:
     lines = self._mapping(x, y, ot.matrix)
     cmap = plt.get_cmap(self._cmap)
     self._lines = []
+
+    # Make the transparency of the lines depend on the strength of the coupling.
+    adjust_transparency = self._adjust_transparency
```
Comment (michalk8) [inline src/ott/tools/plot.py:128]: This code repeats itself below, I'd suggest:
```python
def _get_alpha(self, ot: Transport) -> float:
  if not self._adjust_transparency:
    return self._alpha

  matrix = ot.matrix
  min_matrix = float(jnp.min(matrix))
  max_matrix = float(jnp.max(matrix))
  scaling_factor = float(jnp.max(jnp.array(matrix.shape)))

  if max_matrix == min_matrix:
    return self._alpha
  ...
  return self._alpha * normalized_strength
```

File: src/ott/tools/plot.py [hunk 3]
```diff
@@ -119,14 +123,34 @@ def __call__(self, ot: Transport) -> List[plt.Artist]:
...
+      scaling_factor = float(jnp.max(jnp.array(matrix.shape)))
+
+      if max_matrix == min_matrix:
+        adjust_transparency = False
+
     for start, end, strength in lines:
+
+      if adjust_transparency:
+        normalized_strength = float(strength) / scaling_factor - min_matrix
+        normalized_strength /= 1e-3 + max_matrix - min_matrix
```
Comment (michalk8) [inline src/ott/tools/plot.py:142]: Why the `1e-3`? Based on the above code, you won't get a division by zero anymore.

**Review by michalk8 — CHANGES_REQUESTED** [review]

**Comment by michalk8** [comment]
@gjhuizing any progress on this?

**Comment by gjhuizing** [comment]
Hi, haven't had time yet but I can have a look soon. Is this blocking for something else?

**Comment by michalk8** [comment]
> Hi, haven't had time yet but I can have a look soon. Is this blocking for something else?

Thanks for the info; not pressing at the moment. In #283 , we might add plotting tests after other parts of the code are addressed.

**Comment by gjhuizing** [comment]
okay, it should be a quick thing! will let you know when it's done :)

File: src/ott/tools/plot.py [hunk 4]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
-      result.append((xy[i, [0, 2]], xy[i, [1, 3]], strength))
+
+      # Compute the transparency.
+      if adjust_transparency:
+
+        # make alpha proportional to the coupling strength and between 0 and 1.
+        normalized_strength = (c[i] - min_matrix) / (max_matrix - min_matrix)
+        alpha = self._alpha * float(normalized_strength)
+
+        # Matplotlib's transparency is sensitive to numerical errors.
```
Comment (michalk8) [inline src/ott/tools/plot.py:137]: This this comment can be removed, it's usually assumed `alpha` is in `[0, 1]`.

File: src/ott/tools/plot.py [hunk 5]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
+        alpha = min(1.0, alpha)
```
Comment (michalk8) [inline src/ott/tools/plot.py:138]: Would move outside the `if` + use `np.clip(alpha, 0.0, 1.0)` in case `adjust_transparency` alpha is out of range as well.

File: src/ott/tools/plot.py [hunk 6]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
+    # Loop over the lines and add them to the list.
     for i in range(xy.shape[0]):
+
+      # Line strength is coupling strength times the number of points.
       strength = jnp.max(jnp.array(matrix.shape)) * c[i]
-      result.append((xy[i, [0, 2]], xy[i, [1, 3]], strength))
+
+      # Compute the transparency.
+      if adjust_transparency:
+
```
Comment (michalk8) [inline src/ott/tools/plot.py:132]: Uninformative empty line, remove.

File: src/ott/tools/plot.py [hunk 7]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
+    # We can only adjust transparency if max(c) > min(c).
+    if adjust_transparency:
+      min_matrix, max_matrix = jnp.min(c), jnp.max(c)
+      if max_matrix == min_matrix:
+        adjust_transparency = False
+
+    # Initialize the list of lines of the form (start, end, strength, alpha).
     result = []
+
```
Comment (michalk8) [inline src/ott/tools/plot.py:123]: Uninformative empty line, remove.

File: src/ott/tools/plot.py [hunk 8]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
+    # Loop over the lines and add them to the list.
```
Comment (michalk8) [inline src/ott/tools/plot.py:124]: Uninformative comment, remove.

File: src/ott/tools/plot.py [hunk 9]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
     for i in range(xy.shape[0]):
+
```
Comment (michalk8) [inline src/ott/tools/plot.py:126]: Uninformative empty line, remove.

File: src/ott/tools/plot.py [hunk 10]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
+        # make alpha proportional to the coupling strength and between 0 and 1.
+        normalized_strength = (c[i] - min_matrix) / (max_matrix - min_matrix)
+        alpha = self._alpha * float(normalized_strength)
+
+        # Matplotlib's transparency is sensitive to numerical errors.
+        alpha = min(1.0, alpha)
+      else:
+        alpha = self._alpha
+
+      # Get start and end points, and append the line to the list.
```
Comment (michalk8) [inline src/ott/tools/plot.py:142]: Would remove this comment, the line it's commenting is really simple/intent is obvious.

File: src/ott/tools/plot.py [hunk 11]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
     c = matrix[jnp.where(matrix > self._threshold)]
     xy = jnp.concatenate([x[u], y[v]], axis=-1)
+
+    # Check if we want to adjust transparency.
+    adjust_transparency = self._adjust_transparency
+
+    # We can only adjust transparency if max(c) > min(c).
+    if adjust_transparency:
+      min_matrix, max_matrix = jnp.min(c), jnp.max(c)
+      if max_matrix == min_matrix:
```
Comment (michalk8) [inline src/ott/tools/plot.py:118]: This can be done in one line using `adjust_transparency = min_matrix != max_matrix` instead of the `if` condition.

File: src/ott/tools/plot.py [hunk 12]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
+        adjust_transparency = False
+
+    # Initialize the list of lines of the form (start, end, strength, alpha).
     result = []
+
+    # Loop over the lines and add them to the list.
     for i in range(xy.shape[0]):
+
+      # Line strength is coupling strength times the number of points.
```
Comment (michalk8) [inline src/ott/tools/plot.py:127]: This comment doesn't provide any additional info to the line below, would again remove.

File: src/ott/tools/plot.py [hunk 13]
```diff
@@ -100,13 +104,46 @@ def _scatter(self, ot: Transport):
...
       strength = jnp.max(jnp.array(matrix.shape)) * c[i]
-      result.append((xy[i, [0, 2]], xy[i, [1, 3]], strength))
+
+      # Compute the transparency.
```
Comment (michalk8) [inline src/ott/tools/plot.py:130]: Remove comment.

File: src/ott/tools/plot.py [hunk 14]
```diff
@@ -65,7 +65,9 @@ def __init__(
       cost_threshold: float = -1.0,  # should be negative for animations.
       scale: int = 200,
       show_lines: bool = True,
-      cmap: str = "cool"
+      cmap: str = "cool",
+      adjust_transparency: bool = False,
```
Comment (michalk8) [inline src/ott/tools/plot.py:69]: I think this needs a better name, e.g., `scale_alpha_by_coupling`

**Review by michalk8 — CHANGES_REQUESTED** [review]
Thanks @gjhuizing , just a few minor comments, then it can be merged!
p.s., I fixed the CI in #326, docs linting should be passing again.

File: src/ott/tools/plot.py (hunk 14 above)
Comment (gjhuizing) [reply src/ott/tools/plot.py]: Thanks for the feedback, made the requested changes

**Review by michalk8 — APPROVED** [review]
LGTM @gjhuizing , thanks a lot!
